File: Cauly3.1/MediationSample/CaulyObjectiveCMediationSample/CaulyCustomEvent/CaulyEventBanner.m
//
//  CaulyEventBanner.m
//  CaulyIOSMediationSample
//
//  Created by FSN on 2023/03/15.
//

#import "CaulyEventBanner.h"

@implementation CaulyEventBanner

#pragma mark - Banner Ad Request
- (void)loadBannerForAdConfiguration:(GADMediationBannerAdConfiguration *)adConfiguration completionHandler:(GADMediationBannerLoadCompletionHandler)completionHandler {
    NSLog(@"%s", __PRETTY_FUNCTION__);
    
    __block GADMediationBannerLoadCompletionHandler originalCompletionHandler =
        [completionHandler copy];
    
    _loadCompletionHandler = ^id<GADMediationBannerAdEventDelegate>(
        _Nullable id<GADMediationBannerAd> ad, NSError *_Nullable error) {

      id<GADMediationBannerAdEventDelegate> delegate = nil;
      if (originalCompletionHandler) {
        // Call original handler and hold on to its return value.
        delegate = originalCompletionHandler(ad, error);
      }

      // Release reference to handler. Objects retained by the handler will also be released.
      originalCompletionHandler = nil;

      return delegate;
    };
    
    // admob 대시보드 등록 parameter : cauly appCode
    NSString *adUnit = adConfiguration.credentials.settings[@"parameter"];
    
    CaulyAdSetting * adSetting = [CaulyAdSetting globalSetting];
    
    // 카울리 로그 레벨
    [CaulyAdSetting setLogLevel:CaulyLogLevelTrace];
    
    // iTunes App ID
    adSetting.appId = @"0";
    
    // 카울리 앱 코드
    adSetting.appCode = adUnit;
    
    // 광고 View 크기
    adSetting.adSize = CaulyAdSize_IPhone;
    
    // 화면 전환 효과
    adSetting.animType = CaulyAnimNone;
    
    // 광고 자동 갱신 시간 (기본값)
    adSetting.reloadTime = CaulyReloadTime_30;
    
    // 광고 자동 갱신 사용 여부 (기본값)
    adSetting.useDynamicReloadTime = YES;
    
    // 광고 랜딩 시 WebView 제거 여부
    adSetting.closeOnLanding = YES;
    
    UIViewController *controller = [adConfiguration topViewController];
    adView = [[CaulyAdView alloc] initWithParentViewController:controller];
    [controller.view addSubview:adView];
    
    adView.delegate = self;
    
    // 카울리 배너 광고 요청
    [adView startBannerAdRequest];
}

#pragma mark GADMediationBannerAd implementation

- (nonnull UIView *)view {
    return adView;
}

#pragma - CaulyAdViewDelegate
// 배너 광고 정보 수신 성공
-(void)didReceiveAd:(CaulyAdView *)adView isChargeableAd:(BOOL)isChargeableAd {
    NSLog(@"didReceiveAd");
    // admob에 광고 수신 성공 전달
    _adEventDelegate = _loadCompletionHandler(self, nil);
}

// 배너 광고 정보 수신 실패
- (void)didFailToReceiveAd:(CaulyAdView *)adView errorCode:(int)errorCode errorMsg:(NSString *)errorMsg {
    NSLog(@"didFailToReceiveAd : %d(%@)", errorCode, errorMsg);
    NSError *error = [[NSError alloc] initWithDomain:@"kr.co.cauly.sdk.ios.mediation.sample" code:errorCode userInfo:@{NSLocalizedDescriptionKey:errorMsg}];
    // admob에 광고 수신 실패 전달
    _adEventDelegate = _loadCompletionHandler(nil, error);
}

@end

File: Cauly3.1/MediationSample/CaulySwiftMediationSample/CaulyCustomEvent/CaulyEventBanner.swift
//
//  CaulyEventBanner.swift
//  CaulySwiftMediationSample
//
//  Created by FSN on 2023/07/24.
//

import Foundation
import GoogleMobileAds
import UIKit

class CaulyEventBanner: NSObject, GADMediationBannerAd, CaulyAdViewDelegate {
    /// The Cauly Ad Network banner ad.
    var adView: CaulyAdView?
    
    /// The ad event delegate to forward ad rendering events to the Google Mobile Ads SDK.
    var delegate: GADMediationBannerAdEventDelegate?
    
    /// The ad event delegate to forward ad rendering events to the Google Mobile Ads SDK.
    var completionHandler: GADMediationBannerLoadCompletionHandler?
    
    func loadBanner(for adConfiguration: GADMediationBannerAdConfiguration, completionHandler: @escaping GADMediationBannerLoadCompletionHandler) {
        // Create the bannerView with the appropriate size.
        let adSize = adConfiguration.adSize
        
        // admob 등록 parameter(Cauly 발급 키)를 가져옵니다.
        let adUnit = adConfiguration.credentials.settings["parameter"] as? String
        
        // 상세 설정 항목들은 표 참조, 설정되지 않은 항목들은 기본값으로 설정됩니다.
        let caulySetting = CaulyAdSetting.global()
        CaulyAdSetting.setLogLevel(CaulyLogLevelTrace)   // CaulyLog 레벨
        caulySetting?.appId = "0"                       // App Store 에 등록된 App ID 정보
        caulySetting?.appCode = adUnit                  // admob 등록 parameter (Cauly 발급 키)
        caulySetting?.animType = CaulyAnimNone          // 화면 전환 효과
        caulySetting?.closeOnLanding = true             // App 으로 이동할 때 webview popup 창을 자동으로 닫아줍니다. 기본값을 false
        caulySetting?.useDynamicReloadTime = false
        
        adView = CaulyAdView.init()
        adView?.delegate = self
        
        let frame = CGRect(x: 0, y: 0, width: adSize.size.width, height: adSize.size.height)
        adView?.bounds = frame
        
        self.completionHandler = completionHandler
        adView?.startBannerAdRequest()      // 배너 광고 요청
    }
    
    // MARK: - GADMediationBannerAd implementation
    var view: UIView {
        return adView ?? UIView()
    }
    
    required override init() {
        super.init()
    }
    
    // MARK: - CaulyAdViewDelegate
    // 광고 정보 수신 성공
    func didReceiveAd(_ adView: CaulyAdView!, isChargeableAd: Bool) {
        print("Loaded didReceiveAd callback")
        if let handler = completionHandler {
            delegate = handler(self, nil)
        }
    }
    
    // 광고 정보 수신 실패
    func didFail(toReceiveAd adView: CaulyAdView!, errorCode: Int32, errorMsg: String!) {
        print("didFailToReceiveAd: \(errorCode)(\(errorMsg!)")
        
        let error = NSError(domain: "kr.co.cauly.sdk.ios.mediation.sample", code: Int(errorCode), userInfo: ["description" : errorMsg as Any])
        
        if let handler = completionHandler {
            delegate = handler(nil, error)
        }
    }
    
    // 랜딩 화면 표시
    func willShowLanding(_ adView: CaulyAdView!) {
        print("willShowLandingView")
        delegate?.reportClick()
    }
    
    // 랜딩 화면이 닫혔을 때
    func didCloseLanding(_ adView: CaulyAdView!) {
        print("didCloseLandingView")
    }
}
